src/pqm/pqm.py

```diff
@@ -1,51 +1,402 @@
+    return chi2_contingency(contingency_table)
+
+def _sample_reference_indices_torch(Nx, nx, Ny, ny, Ng, x_samples, y_samples, device):
+    """
+    Helper function to sample references for GPU-based Torch computations.
+
+    Parameters
+    ----------
+    Nx : int
+        Number of references to sample from x_samples.
+    nx : int
+        Number of samples in x_samples.
+    Ny : int
+        Number of references to sample from y_samples.
+    ny : int
+        Number of samples in y_samples.
+    Ng : int
+        Number of references to sample from a Gaussian distribution.
+
+    Returns
+    -------
+    np.ndarray  
+        References samples.
+    """
+    
+    if Nx > nx:
+        raise ValueError("Cannot sample more references from x_samples than available")
+    if Ny > ny:
+        raise ValueError("Cannot sample more references from y_samples than available")
+    
+    # Reference samples from x_samples
+    x_indices = torch.randperm(nx, device=device)
+    xrefs_indices = x_indices[:Nx]
+    x_samples_indices = x_indices[Nx:]
+    xrefs = x_samples[xrefs_indices]
+    x_samples = x_samples[x_samples_indices]
+    
+    # Reference samples from y_samples
+    y_indices = torch.randperm(ny, device=device)
+    yrefs_indices = y_indices[:Ny]
+    y_samples_indices = y_indices[Ny:]
+    yrefs = y_samples[yrefs_indices]
+    y_samples = y_samples[y_samples_indices]
+    
+    # Combine references
+    refs = torch.cat([xrefs, yrefs], dim=0)
+    
+    # Gaussian references
+    if Ng > 0:
+        m, s = _mean_std(x_samples, y_samples)
+        # Ensure m and s have the correct shape
+        if m.dim() == 1:
+            m = m.unsqueeze(0)
+        if s.dim() == 1:
+            s = s.unsqueeze(0)
+        gauss_refs = torch.normal(
+            mean=m.repeat(Ng, 1),
+            std=s.repeat(Ng, 1),
+        )
+        refs = torch.cat([refs, gauss_refs], dim=0)
+    return refs
+    
+def _compute_distances_torch(x_samples, y_samples, refs, current_num_refs, num_refs):
+    """
+    Helper function to calculate distances for GPU-based Torch computations.
+
+    Parameters
+    ----------
+    x_samples : torch.Tensor
+        Samples from the first distribution. Must have shape (N, *D) N is the
+        number of x samples, and D is the dimensionality of the samples.
+    y_samples : torch.Tensor
+        Samples from the second distribution. Must have shape (M, *D) M is the
+        number of y samples, and D is the dimensionality of the samples.
+    refs : torch.Tensor
+        Reference samples. Must have shape (num_refs, *D) where D is the
+        dimensionality of the samples.
+    current_num_refs : int
+        Number of reference samples used in the test.
+    num_refs : int
+        Number of reference samples to use.
+
+    Returns
+    -------
+    tuple
+        Results from the PyTorch implementation of chi2_contingency.
+    """
+
+    # Compute distances and find nearest references
```

Comment:
Should rearrange either this one or the numpy one so that they are organized the same way. This does all x then all y, whereas the numpy one does both cdist, both argmin, both bincount.